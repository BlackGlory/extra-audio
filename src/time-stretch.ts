import { getSample } from '@utils/get-sample.js'

export function timeStretch(
  input: Float32Array
, output: Float32Array
): Float32Array {
  // outputIndex / inputIndex = outputLength / inputLength
  const scale = output.length / input.length

  for (let outputIndex = output.length; outputIndex--;) {
    const inputIndex = outputIndex / scale
    output[outputIndex] = getSample(input, inputIndex)
  }

  return output
}
