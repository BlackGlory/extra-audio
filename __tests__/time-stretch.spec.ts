import { describe, it, expect } from 'vitest'
import { timeStretch } from '@src/time-stretch.js'

describe('timeStretch', () => {
  it('output.length === input.length', () => {
    const input = new Float32Array([0, 1])
    const output = new Float32Array(input.length)

    const result = timeStretch(input, output)

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(input)
  })

  it('output.length > input.length', () => {
    const input = new Float32Array([1, 2])
    const output = new Float32Array(4)

    const result = timeStretch(input, output)

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(new Float32Array([1, 1.5, 2, 2]))
  })

  it('output.length < input.length', () => {
    const input = new Float32Array([1, 1.5, 2, 2])
    const output = new Float32Array(input.length / 2)

    const result = timeStretch(input, output)

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(new Float32Array([1, 2]))
  })
})
