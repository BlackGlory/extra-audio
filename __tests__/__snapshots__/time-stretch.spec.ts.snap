// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`timeStretch > output.length < input.length 1`] = `
Float32Array [
  1,
  2,
]
`;

exports[`timeStretch > output.length > input.length 1`] = `
Float32Array [
  1,
  1.5,
  2,
  2,
]
`;
